idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/components/KtExpressionInfoProvider.kt

```diff
@@ -7,17 +7,30 @@ package org.jetbrains.kotlin.idea.frontend.api.components
 
 import org.jetbrains.kotlin.diagnostics.WhenMissingCase
 import org.jetbrains.kotlin.idea.frontend.api.symbols.KtCallableSymbol
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtValueParameterSymbol
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtLambdaExpression
 import org.jetbrains.kotlin.psi.KtReturnExpression
 import org.jetbrains.kotlin.psi.KtWhenExpression
 
 public abstract class KtExpressionInfoProvider : KtAnalysisSessionComponent() {
     public abstract fun getReturnExpressionTargetSymbol(returnExpression: KtReturnExpression): KtCallableSymbol?
     public abstract fun getWhenMissingCases(whenExpression: KtWhenExpression): List<WhenMissingCase>
+
+    public abstract fun getLastExpressionIfImplicitReturn(ktLambdaExpression: KtLambdaExpression): KtExpression?
+    public abstract fun getParameters(ktLambdaExpression: KtLambdaExpression): List<KtValueParameterSymbol>
 }
 
 public interface KtExpressionInfoProviderMixIn : KtAnalysisSessionMixIn {
     public fun KtReturnExpression.getReturnTargetSymbol(): KtCallableSymbol? =
         analysisSession.expressionInfoProvider.getReturnExpressionTargetSymbol(this)
 
-    public fun KtWhenExpression.getMissingCases(): List<WhenMissingCase> = analysisSession.expressionInfoProvider.getWhenMissingCases(this)
-}
\ No newline at end of file
+    public fun KtWhenExpression.getMissingCases(): List<WhenMissingCase> =
+        analysisSession.expressionInfoProvider.getWhenMissingCases(this)
+
+    public fun KtLambdaExpression.getLastExpressionIfImplicitReturn(): KtExpression? =
+        analysisSession.expressionInfoProvider.getLastExpressionIfImplicitReturn(this)
+
+    public fun KtLambdaExpression.getParameters(): List<KtValueParameterSymbol> =
```

Comment:
This function seems redundant. You can create `KtAnonymousFunctionSymbol` in your UAST code by `KtFunctionLiteral` and then get the list of parameters:
```
analyseForUast(context) {
   val parameters = ktLambdaExpression.functionLiteral.getAnonymousFunctionSymbol().valueParameters
}
```

Reply:
Using `KtFunctionLiteral` as a key doesn't work: it simply fails at `withFirDeclarationOfType`. Instead, using its parent, `KtLambdaExpression` works. Added a test to showcase an example from UAST. After that, yes, value parameters can be retrieved from anonymous function symbol.